File: android/app/src/main/java/com/therealdeltrex/localeventfinder/data/Places.kt
package com.therealdeltrex.localeventfinder.data

import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.selects.select
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume

/**
 * Permanent "things to do" from OpenStreetMap via the Overpass API — museums,
 * parks, zoos, galleries, cinemas, castles, mini-golf, beaches and so on.
 * Ported from places.py.
 *
 * Uses `nwr` (nodes + ways + relations) because many attractions are mapped as
 * areas, not points (e.g. the Rømø labyrinth). The three types are output
 * separately so a single result cap can't drop all the areas. The public
 * servers are slow/flaky, so all mirrors are queried in parallel and the first
 * to answer wins.
 */
class Places(client: OkHttpClient) {

    // Overpass can take ~30–40s; give these calls their own generous timeouts.
    private val client = client.newBuilder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .callTimeout(60, TimeUnit.SECONDS)
        .build()

    private val endpoints = listOf(
        "https://overpass-api.de/api/interpreter",
        "https://overpass.kumi.systems/api/interpreter",
        "https://overpass.private.coffee/api/interpreter",
        "https://maps.mail.ru/osm/tools/overpass/api/interpreter",
    )

    private val keys = listOf("tourism", "leisure", "amenity", "historic", "natural")

    // "key|value" -> (label, tags)
    private val cats: Map<String, Pair<String, List<String>>> = linkedMapOf(
        "tourism|museum" to ("Museum" to listOf(Tagging.TAG_FAMILY)),
        "tourism|gallery" to ("Gallery" to listOf(Tagging.TAG_DATE)),
        "tourism|zoo" to ("Zoo" to listOf(Tagging.TAG_FAMILY)),
        "tourism|theme_park" to ("Theme park" to listOf(Tagging.TAG_FAMILY)),
        "tourism|aquarium" to ("Aquarium" to listOf(Tagging.TAG_FAMILY)),
        "tourism|attraction" to ("Attraction" to emptyList()),
        "tourism|viewpoint" to ("Viewpoint" to listOf(Tagging.TAG_DATE)),
        "tourism|artwork" to ("Public art" to listOf(Tagging.TAG_DATE)),
        "leisure|park" to ("Park" to listOf(Tagging.TAG_FAMILY)),
        "leisure|garden" to ("Garden" to listOf(Tagging.TAG_DATE)),
        "leisure|nature_reserve" to ("Nature reserve" to emptyList()),
        "leisure|water_park" to ("Water park" to listOf(Tagging.TAG_FAMILY)),
        "leisure|miniature_golf" to ("Mini golf" to listOf(Tagging.TAG_FAMILY)),
        "leisure|golf_course" to ("Golf course" to emptyList()),
        "leisure|sports_centre" to ("Sports centre" to listOf(Tagging.TAG_FAMILY)),
        "leisure|swimming_pool" to ("Swimming pool" to listOf(Tagging.TAG_FAMILY)),
        "leisure|bathing_place" to ("Bathing spot" to listOf(Tagging.TAG_FAMILY)),
        "leisure|horse_riding" to ("Horse riding" to listOf(Tagging.TAG_FAMILY)),
        "leisure|playground" to ("Playground" to listOf(Tagging.TAG_FAMILY)),
        "leisure|bird_hide" to ("Bird hide" to listOf(Tagging.TAG_DATE)),
        "amenity|cinema" to ("Cinema" to listOf(Tagging.TAG_DATE)),
        "amenity|theatre" to ("Theatre" to listOf(Tagging.TAG_DATE)),
        "amenity|arts_centre" to ("Arts centre" to listOf(Tagging.TAG_DATE)),
        "amenity|public_bath" to ("Swimming pool" to listOf(Tagging.TAG_FAMILY)),
        "historic|castle" to ("Castle" to listOf(Tagging.TAG_DATE)),
        "historic|monument" to ("Monument" to emptyList()),
        "historic|memorial" to ("Memorial" to emptyList()),
        "historic|ruins" to ("Ruins" to listOf(Tagging.TAG_DATE)),
        "natural|beach" to ("Beach" to listOf(Tagging.TAG_FAMILY)),
    )

    private fun query(lat: Double, lon: Double, radiusM: Int): String {
        val byKey = LinkedHashMap<String, MutableList<String>>()
        for (k in cats.keys) {
            val (key, value) = k.split("|")
            byKey.getOrPut(key) { mutableListOf() }.add(value)
        }
        val body = byKey.entries.joinToString("") { (key, vals) ->
            "nwr(around:$radiusM,$lat,$lon)[\"$key\"~\"^(${vals.joinToString("|")})$\"];"
        }
        return "[out:json][timeout:60];($body)->.a;" +
            "node.a;out 300;way.a;out center 300;relation.a;out center 100;"
    }

    private fun categoryOf(tags: JSONObject): Pair<String, List<String>>? {
        // Public pools are tagged inconsistently — label any of them "Swimming pool".
        val sport = tags.optString("sport", "").lowercase()
        val leisure = tags.optString("leisure", "")
        if (sport.contains("swimming") || leisure == "swimming_pool" ||
            leisure == "water_park" || tags.optString("amenity", "") == "public_bath") {
            return "Swimming pool" to listOf(Tagging.TAG_FAMILY)
        }
        for (key in keys) {
            val v = tags.optString(key, "")
            if (v.isNotEmpty()) cats["$key|$v"]?.let { return it }
        }
        return null
    }

    /** POST to one mirror; returns the elements array, or null on any failure. */
    private suspend fun queryOne(endpoint: String, body: String): JSONArray? =
        suspendCancellableCoroutine { cont ->
            val req = Request.Builder()
                .url(endpoint)
                .post(FormBody.Builder().add("data", body).build())
                .header("User-Agent", "LocalEventsFinder-Android/1.0")
                .build()
            val call = client.newCall(req)
            cont.invokeOnCancellation { runCatching { call.cancel() } }
            call.enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    if (cont.isActive) cont.resume(null)
                }

                override fun onResponse(call: Call, response: Response) {
                    val arr = response.use {
                        try {
                            if (it.code == 200)
                                JSONObject(it.body?.string() ?: "").optJSONArray("elements")
                            else null
                        } catch (e: Exception) {
                            null
                        }
                    }
                    if (cont.isActive) cont.resume(arr)
                }
            })
        }

    /** Race all mirrors; return the first non-null response, cancel the rest. */
    private suspend fun race(body: String): JSONArray? = coroutineScope {
        val deferreds = endpoints.map { ep -> async { queryOne(ep, body) } }
        var result: JSONArray? = null
        val remaining = deferreds.toMutableList()
        while (remaining.isNotEmpty() && result == null) {
            val (winner, res) = select<Pair<Deferred<JSONArray?>, JSONArray?>> {
                remaining.forEach { d -> d.onAwait { d to it } }
            }
            remaining.remove(winner)
            if (res != null) result = res
        }
        deferreds.forEach { it.cancel() }
        result
    }

    suspend fun fetch(lat: Double, lon: Double, radiusKm: Double, limit: Int = 120): List<Event> {
        val elements = race(query(lat, lon, (radiusKm * 1000).toInt())) ?: return emptyList()

        val seen = HashSet<String>()
        val out = ArrayList<Event>()
        for (i in 0 until elements.length()) {
            val el = elements.optJSONObject(i) ?: continue
            val tags = el.optJSONObject("tags") ?: continue
            val name = tags.optString("name", "")
            if (name.isEmpty()) continue
            val cat = categoryOf(tags) ?: continue
            val center = el.optJSONObject("center")
            val plat = if (el.has("lat")) el.optDouble("lat") else center?.optDouble("lat") ?: continue
            val plon = if (el.has("lon")) el.optDouble("lon") else center?.optDouble("lon") ?: continue
            // dedupe by name + coarse location (~1 km) so a twice-mapped place collapses
            val key = "${name.lowercase()}|${"%.2f".format(plat)}|${"%.2f".format(plon)}"
            if (!seen.add(key)) continue
            val osmUrl = "https://www.openstreetmap.org/${el.optString("type")}/${el.optString("id")}"
            val website = tags.optString("website").ifEmpty {
                tags.optString("contact:website").ifEmpty { osmUrl }
            }
            out.add(
                Event(
                    id = "poi-${el.optString("type")}-${el.optString("id")}",
                    title = name,
                    url = website,
                    source = "OpenStreetMap",
                    description = cat.first,
                    permanent = true,
                    category = cat.first,
                    hours = tags.optString("opening_hours", ""),
                    locality = tags.optString("addr:city", ""),
                    lat = plat,
                    lon = plon,
                    autoTags = cat.second.toSet(),
                )
            )
        }
        // nearest first, then cap — so a close area is never truncated away
        out.sortBy { Geo.haversineKm(lat, lon, it.lat!!, it.lon!!) }
        return out.take(limit)
    }
}

File: eventfinder/places.py
"""Permanent 'things to do' from OpenStreetMap via the Overpass API.

Scraped event feeds only cover scheduled happenings, and in a given area there
may be few of those. This adds the places you can visit *any* day — museums,
parks, zoos, galleries, cinemas, castles and so on — so the finder always has
something to show within range. No API key needed.
"""
from __future__ import annotations

import concurrent.futures
from typing import Optional

import requests

from . import geo
from .models import Event, event_id

# The public Overpass instances get overloaded and return 504/429; rotate
# through mirrors until one answers with valid JSON.
OVERPASS_ENDPOINTS = [
    "https://overpass-api.de/api/interpreter",
    "https://overpass.kumi.systems/api/interpreter",
    "https://overpass.private.coffee/api/interpreter",
    "https://maps.mail.ru/osm/tools/overpass/api/interpreter",
]

# OSM key order to check, most specific first.
_KEYS = ("tourism", "leisure", "amenity", "historic", "natural")

# OSM (key, value) -> (human label, tags). Drives both the query and tagging.
CATEGORIES: dict[tuple[str, str], tuple[str, list[str]]] = {
    ("tourism", "museum"): ("Museum", ["family"]),
    ("tourism", "gallery"): ("Gallery", ["date"]),
    ("tourism", "zoo"): ("Zoo", ["family"]),
    ("tourism", "theme_park"): ("Theme park", ["family"]),
    ("tourism", "aquarium"): ("Aquarium", ["family"]),
    ("tourism", "attraction"): ("Attraction", []),
    ("tourism", "viewpoint"): ("Viewpoint", ["date"]),
    ("tourism", "artwork"): ("Public art", ["date"]),
    ("leisure", "park"): ("Park", ["family"]),
    ("leisure", "garden"): ("Garden", ["date"]),
    ("leisure", "nature_reserve"): ("Nature reserve", []),
    ("leisure", "water_park"): ("Water park", ["family"]),
    ("leisure", "miniature_golf"): ("Mini golf", ["family"]),
    ("leisure", "golf_course"): ("Golf course", []),
    ("leisure", "sports_centre"): ("Sports centre", ["family"]),
    ("leisure", "swimming_pool"): ("Swimming pool", ["family"]),
    ("leisure", "bathing_place"): ("Bathing spot", ["family"]),
    ("leisure", "horse_riding"): ("Horse riding", ["family"]),
    ("leisure", "playground"): ("Playground", ["family"]),
    ("leisure", "bird_hide"): ("Bird hide", ["date"]),
    ("amenity", "cinema"): ("Cinema", ["date"]),
    ("amenity", "theatre"): ("Theatre", ["date"]),
    ("amenity", "arts_centre"): ("Arts centre", ["date"]),
    ("amenity", "public_bath"): ("Swimming pool", ["family"]),
    ("historic", "castle"): ("Castle", ["date"]),
    ("historic", "monument"): ("Monument", []),
    ("historic", "memorial"): ("Memorial", []),
    ("historic", "ruins"): ("Ruins", ["date"]),
    ("natural", "beach"): ("Beach", ["family"]),
}


def _values(key: str) -> str:
    vals = [v for (k, v) in CATEGORIES if k == key]
    return "|".join(vals)


def build_query(lat: float, lon: float, radius_m: int) -> str:
    # nwr = nodes + ways + relations. Many attractions (parks, museums, the
    # Rømø labyrinth) are mapped as areas, not points, so a nodes-only query
    # misses them. `out center` gives each area a representative coordinate.
    clauses = []
    for key in _KEYS:
        clauses.append(
            f'nwr(around:{radius_m},{lat},{lon})["{key}"~"^({_values(key)})$"];'
        )
    body = "".join(clauses)
    # Output nodes, ways and relations *separately* — a single `out` emits all
    # nodes first, so a plain cap would exhaust its slots on nodes and drop
    # every area (the labyrinth is a way). Per-type caps guarantee areas appear.
    return (
        f"[out:json][timeout:60];({body})->.a;"
        f"node.a;out 300;way.a;out center 300;relation.a;out center 100;"
    )


def _category_of(tags: dict) -> Optional[tuple[str, list[str]]]:
    # Public swimming pools are mapped inconsistently (leisure=swimming_pool,
    # water_park, or sports_centre+sport=swimming, amenity=public_bath). Detect
    # any of them and label clearly as "Swimming pool".
    sport = (tags.get("sport") or "").lower()
    if (
        "swimming" in sport
        or tags.get("leisure") in ("swimming_pool", "water_park")
        or tags.get("amenity") == "public_bath"
    ):
        return ("Swimming pool", ["family"])
    for key in _KEYS:
        val = tags.get(key)
        if val and (key, val) in CATEGORIES:
            return CATEGORIES[(key, val)]
    return None


def _query_one(endpoint: str, query: str) -> Optional[list]:
    try:
        resp = requests.post(
            endpoint,
            data={"data": query},
            headers={"User-Agent": geo.USER_AGENT},
            timeout=55,
        )
        if resp.status_code == 200:
            return resp.json().get("elements", [])
    except (requests.RequestException, ValueError):
        return None
    return None


def run_overpass(query: str) -> Optional[list]:
    """Run an Overpass query against all mirrors in parallel, returning the
    elements from the first that answers (or None)."""
    executor = concurrent.futures.ThreadPoolExecutor(max_workers=len(OVERPASS_ENDPOINTS))
    futures = [executor.submit(_query_one, ep, query) for ep in OVERPASS_ENDPOINTS]
    try:
        for fut in concurrent.futures.as_completed(futures):
            res = fut.result()
            if res is not None:
                return res
    finally:
        executor.shutdown(wait=False)  # don't block on the losing mirrors
    return None


def nearby_localities(lat: float, lon: float, radius_km: float, limit: int = 8) -> list[str]:
    """Nearest named towns/cities within the radius — used to look up events in
    neighbouring places, not just the origin's own (which may have none)."""
    radius_m = int(radius_km * 1000)
    # `out` (body) gives coordinates + tags for nodes; `out tags` omits coords.
    query = (
        f'[out:json][timeout:40];'
        f'(node(around:{radius_m},{lat},{lon})[place~"^(city|town)$"];);out 200;'
    )
    elements = run_overpass(query) or []
    scored: list[tuple[float, str]] = []
    for el in elements:
        name = (el.get("tags") or {}).get("name")
        plat, plon = el.get("lat"), el.get("lon")
        if name and plat is not None and plon is not None:
            scored.append((geo.haversine_km(lat, lon, plat, plon), name))
    scored.sort()
    seen: set[str] = set()
    names: list[str] = []
    for _, name in scored:
        key = name.lower()
        if key not in seen:
            seen.add(key)
            names.append(name)
    return names[:limit]


def fetch_places(lat: float, lon: float, radius_km: float, limit: int = 120) -> list[Event]:
    """Return named permanent places within radius.

    The public Overpass servers are individually slow and unreliable, so we
    hit all mirrors in parallel and take the first that answers — bounding the
    wait to the fastest server instead of the sum of their timeouts.
    """
    radius_m = int(radius_km * 1000)
    elements = run_overpass(build_query(lat, lon, radius_m))
    if elements is None:
        return []

    seen: set[str] = set()
    out: list[Event] = []
    for el in elements:
        tags = el.get("tags") or {}
        name = tags.get("name")
        if not name:
            continue
        cat = _category_of(tags)
        if not cat:
            continue
        label, cat_tags = cat
        plat = el.get("lat") or (el.get("center") or {}).get("lat")
        plon = el.get("lon") or (el.get("center") or {}).get("lon")
        if plat is None or plon is None:
            continue
        # Dedupe by name + coarse location (~1 km) so a facility mapped twice
        # (e.g. the pool basin and the building) collapses into one result.
        key = f"{name.lower()}|{round(float(plat), 2)}|{round(float(plon), 2)}"
        if key in seen:
            continue
        seen.add(key)
        osm_url = f"https://www.openstreetmap.org/{el.get('type')}/{el.get('id')}"
        website = tags.get("website") or tags.get("contact:website") or osm_url
        out.append(
            Event(
                id=event_id(osm_url, name, label),
                title=name,
                url=website,
                source="OpenStreetMap",
                description=label,
                permanent=True,
                category=label,
                locality=tags.get("addr:city", ""),
                lat=float(plat),
                lon=float(plon),
                hours=tags.get("opening_hours", ""),
                auto_tags=list(cat_tags),
            )
        )
    # Overpass returns by id, not distance; keep the nearest `limit` so a close
    # place (e.g. the labyrinth around the corner) is never truncated away.
    out.sort(key=lambda e: geo.haversine_km(lat, lon, e.lat, e.lon))
    return out[: max(limit, 0)] if limit else out
